tracker/enqueue_test.go

```diff
@@ -380,3 +380,261 @@ func TestBadObjectReferences(t *testing.T) {
+		name   string
+		objRef Reference
+		match  string
+	}{{
+		name:   "Missing All",
+		objRef: Reference{},
+		match:  "\nAPIVersion:.*\nKind:.*\nNamespace:",
+	}, {
+		name: "Name and Selector",
+		objRef: Reference{
+			APIVersion: "apps/v1",
+			Kind:       "Deployment",
+			Namespace:  "default",
+			Name:       "foo",
+			Selector:   &metav1.LabelSelector{},
+		},
+		match: "both Name and Selector",
+	}, {
+		name: "bad label key",
+		objRef: Reference{
+			APIVersion: "apps/v1",
+			Kind:       "Deployment",
+			Namespace:  "default",
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"a bad key": "bar",
+				},
+			},
+		},
+		match: "a bad key",
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := trk.TrackReference(test.objRef, thing1); err == nil {
+				t.Error("Track() = nil, wanted error")
+			} else {
+				match, e2 := regexp.Match(test.match, []byte(err.Error()))
+				if e2 != nil {
+					t.Errorf("Failed to compile %q: %v", e2, test.match)
+				} else if !match {
+					t.Errorf("Track() = %v, wanted match: %s", err, test.match)
+				}
+			}
+		})
+	}
+}
+
+func TestHappyPathsInexact(t *testing.T) {
+	calls := 0
+	f := func(key types.NamespacedName) {
+		calls = calls + 1
+	}
+
+	trk := New(f, 100*time.Millisecond)
+
+	thing1 := &Resource{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "ref.knative.dev/v1alpha1",
+			Kind:       "Thing1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "ns",
+			Name:      "foo",
+			Labels: map[string]string{
+				"foo": "bar",
+				// An extra label.
+				"baz": "blah",
+			},
+		},
+	}
+	or := kmeta.ObjectReference(thing1)
+	ref := Reference{
+		APIVersion: or.APIVersion,
+		Kind:       or.Kind,
+		Namespace:  or.Namespace,
+		Selector: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"foo": "bar",
+			},
+		},
+	}
+
+	thing2 := &Resource{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "reffer.knative.dev/v1alpha1",
+			Kind:       "Thing2",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "default",
+			Name:      "bar.baz.this-is-fine",
+		},
+	}
+
+	// Not tracked yet
+	{
+		trk.OnChanged(thing1)
+		if got, want := calls, 0; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Tracked gets called
+	{
+		if err := trk.TrackReference(ref, thing2); err != nil {
+			t.Errorf("Track() = %v", err)
+		}
+		// New registrations should result in an immediate callback.
+		if got, want := calls, 1; got != want {
+			t.Errorf("Track() = %v, wanted %v", got, want)
+		}
+
+		trk.OnChanged(thing1)
+		if got, want := calls, 2; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Still gets called
+	{
+		trk.OnChanged(thing1)
+		if got, want := calls, 3; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Check that after the sleep duration, we stop getting called.
+	time.Sleep(200 * time.Millisecond)
```

Comment:
Your timeout is 100. Sleeping 101 should suffice?

Reply:
Sure...